core/src/main/scala/kafka/log/Log.scala

```diff
@@ -306,86 +306,136 @@ class Log(@volatile var dir: File,
 
   def highWatermark: Long = highWatermarkMetadata.messageOffset
 
-  def highWatermark_=(newHighWatermark: Long): Unit = {
-    highWatermarkMetadata = LogOffsetMetadata(newHighWatermark)
+  /**
+   * Update the high watermark to a new offset. The new high watermark will be lower
+   * bounded by the log start offset and upper bounded by the log end offset.
+   *
+   * This is intended to be called when initializing the high watermark or when updating
+   * it on a follower after receiving a Fetch response from the leader.
+   *
+   * @param hw the suggested new value for the high watermark
+   * @return the updated high watermark offset
+   */
+  def updateHighWatermark(hw: Long): Long = {
```

Comment:
Should this be named `initializeHighWatermark` maybe?

Reply:
I called it `update` since it is also used in the fetcher when updating the high watermark after a fetch response.